### ISSUE TITLE
Webtoon view lag with large stripes

### ISSUE THREAD

**Issue by whiterblue** [inorichi/tachiyomi#1908]
**App version:** 0.8.1

**Issue/Request:** The app makes the whole phone lag heavily when using the webtoon view mode with webtoon image stripes that are very long and with high resolution (mangadex case). Makes sense that it lags at these extreme cases but would be nice not to.
Thanks. 

**Related:** #1521 

**Steps to reproduce:**
 1. Default settings.
 2. Set image view as webtoon.
 3. Search for "skeleton soldier ... " in  Mangadex.
 4. Load last chapter (56 in my case).
 5. Scroll and see the phone lag 

**Extra info:**
- My smartphone is not super powerful but can load any webpage that you throw at it so this should not lag.
- 2gb of the 3gb of ram is used when the lag happens (normal in my phone)
- No background apps doing heavy tasks

**Comment by ZaneHannanAU** [comment]
Same here, app version r1471, though different manga.
https://mangadex.org/manga/34587/
I think a means we would have is to reslice where possible, but that would likely end up under advanced options.

Would also go for an optimize option (eg jpg -> webp 80% lossy, png -> webp lossless, webp -> identity, gif -> webp if smaller).
Would add a subsubmenu with some options, eg "reduce size when above XxY" and so on.

**Comment by ZaneHannanAU** [comment]
My phone is a nexus 6P, for reference.

**Comment by inorichi** [comment]
It doesn't matter that you compress the image with lower quality. When it's decoded it has the same size on memory.

The only proper solution is to split the images so the view recycling works better, and that's not easy on the fly...

**Comment by ZaneHannanAU** [comment]
I was going to say that "reduce size when above XxY" would be done on download. Would also open up a thing to add optimization sliders.
So like reslice the images as a post-processing detail?

**Comment by inorichi** [comment]
Yes, but I don't think it'd be easy to implement.

**Comment by ZaneHannanAU** [comment]
me neither
I was considering making it a library function for my ripster project for webtoons but never ended up doing it

It was basically checking that 15 rows had within ~1% of the same colour (abs(rgba1 - rgba2) < (3,3,3,0)) and just sliced them as possible.

The hard part is reencoding, but with webp on lossless and low memory it goes decently fast.

But it also required stitching the entire image together to break it apart afterwards.

**Comment by inorichi** [comment]
Wait, that's not what I had in mind.

I was talking about splitting the images into several, smaller portions. The problem is that we are given a fixed amount of "pages" and I'd need a RecyclerView of RecyclerViews to load the regions, which is not easy.

**Comment by inorichi** [comment]
Let me explain the problem better.

When a page is 2000x20000, the subsampling library will load them in 10 pages of 2000x2000. The problem is that the RecyclerView will bind the entire page and load all 10 pages at once, even if they are outside the screen.

And it gets worse when we consider that there might be more view holders bound (I think with these sizes only 3 are loaded), so we'd end up with 30 pages of 2000x2000. As they are decoded with RGB565, this means we have 30*2000*2000*2 = 228MB in RAM.

But the problem is in the RenderThread that starts lagging with such huge bitmaps.

I'd like to know a way to bind a particular region of a RecyclerView, but it was not designed to display these big images so it's not his fault. The only solution I see is either a custom LayoutManager (not sure it'd help) or a custom view. Both are hell.

**Comment by ZaneHannanAU** [comment]
… oof

**Comment by ZaneHannanAU** [comment]
… this makes me glad I only deal with recycling a few (3--5) canvases and just move them around in my prototypes … and I reslice just about everything anyway …

**Comment by e-shl** [comment]
May help increase stability https://github.com/bloc97/Anime4K

**Comment by derencel20** [comment]
![Screenshot_20200911_001820](https://user-images.githubusercontent.com/17394978/92761514-59907380-f3c4-11ea-8508-0ce7e02903ef.jpg)
disabling this solved the problem for me. Idk about others tho.

**Comment by leumasme** [comment]
Noticing this as well, however only on specific parts of the strip (on page seams?).
Lagging just as badly on my new A50 as on my old A3.
32bit color is disabled.

**Comment by Lightjohn** [comment]
I have mostly the same issue: pretty big images and old harware and I went for side application I made that will slice big webtoon images in smaller of height equivalent of screen size.

For example: going from one image 6000x20000 (000.jpg) to 12 images 6000x1800 (000-001.pg 000-002.jpg) and after that no more issue, reader is quite snappy and memory usge is quite low.

The code for that in my case is just ~15 lines and does not need anything from Tachiyomi at this point.
Slicing big images seems to be ths solution for me.

I could open source my application if people are interested.

**Comment by arkon** [comment]
@Lightjohn Yeah that sounds nice. You could consider submitting a PR for that feature as well.

**Comment by Lightjohn** [comment]
I could do it but I know nothing about Tachiyomi infrastructure which is why I went for my own Android application.

If you could guide me on where to look, will try to give a go.

**Comment by arkon** [comment]
This should be a good entry point: https://github.com/inorichi/tachiyomi/blob/dev/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt#L269

**Comment by Lightjohn** [comment]
Thanks:
1/ do we want that feature enabled by default/everytime?
 2/ Should I keep my idea to split when height is > screenHeight, it could affect some normal big images.
Maybe depending of image ratio?
Never done kotlin but seems close enough to java for me to try it.

**Comment by adityaravishankar** [comment]
The way I see it, the reader works really well when webtoon pages are tiny strips. The loader loads them easily and places them below faster. So even if you were to split images to half the screen height, the renderer could load the images faster without preloading excess data in memory. 

Also since splitting only makes sense for webtoons, and may be extra processing that is not needed for all webtoons, I'd recommend a toggle that is disabled by default and enabled per comic. 

I'm going to try and set up the build environment today and see if I can figure out a clean and simple fix. Do keep sharing your progress as well.

**Comment by Lightjohn** [comment]
You are right: 99% of the current webtoons (which is already already a subset of all the things that can be read with tachiyomi) are working perfectly.
And for now I only have two sources that does that.

The "perfect" fix would be only a partial image load / display but for what I read in this thread it's not that easy.

What I will do: I will implement my solution it's easy and simple but I will leave the triggering condition to you :)

**Comment by inorichi** [comment]
I'm starting to look into this on 1.x with compose, and it's promising. So, the best solution is to retrieve the image size with the region decoder, and then calculate how many tiles we'll need. Next, we replace that page with the calculated amount of page tiles. That way, we still use a single RecyclerView (LazyColumn on compose) with sane sizes and proper recycling.

However, I assume it'd be a big refactor of the webtoon viewer and we'd also need to integrate the border cropping library outside SubsamplingScaleImageView because the webtoon viewer has to manage the image size and tiles itself.

**Comment by arkon** [comment]
From #5945, OOM happens in other modes too depending on scaling:

> This issues also hapens when choosing the option to fit width in the other modes

**Comment by Lightjohn** [comment]
In the meantime, for people that have issue with reading big images I wrote a small application that will split images:
I use it everyday because I don't have a lot of ram and some webtoon are now in 2-5 images...

https://github.com/Lightjohn/TachiyomiImageSplitter

**Comment by lemon07r** [comment]
I still have this issue, more noticably on some webtoons. For example, right now I'm having bad stutter when scrolling through chapter 1 of the legendary mechanic (source: mangakalot), in webtoon view. 

My phone is a Xiaomi k40 pro on the 12.5.7 stable xiaomi.eu ROM (this is based on the performance edition Chinese ROM). 

Can't be a hardware issue, this is a top end performance phone with a 120hz screen. Are there any options or settings I can adjust to fix this issue?

**Comment by Krilluz** [comment]
Dont know if relevant, but the in some webtoons or certain pages of one, if you open the long press menu while in the seccion/s with the issue, the lag in that specific part is gone. it can also work activating the cut borders option, again, while in the lagging section.
It actualy worked 100% of the time, like 2 months ago ?,but  now as the issue  became more frecuent in a lot more webtoons, it only works in 1/3 of the cases.

https://user-images.githubusercontent.com/98551443/153107129-fccb4556-f0a8-4813-aac6-b8c2b168dc2a.mp4

**Comment by arkon** [comment]
As noted in #6973, a potential solution would be to only split downloaded images.

**Comment by X9VoiD** [comment]
Would like to add that this issue also exists on **non-webtoon** images

**Steps to reproduce:**
1. Install Mangadex extension.
2. Browse to [Isekai Monster Breeder](https://mangadex.org/title/0a89a48d-2e94-49a7-aa05-61b24e0f2f86/isekai-monster-breeder).
3. Open Chapter 45 scanlated by Emily Manga Translations.
4. Navigate to Page 2. This and every page after page 2 has very large dimensions and lags during panning. Even just bringing up the UI by tapping once anywhere in the middle is also laggy.

This is reproducible even on an LG G7 which has a relatively strong GPU. *Reading mode doesn't matter.*

The solution in #7010 first proposed by #6973 doesn't fix this and the underlying problem. Furthermore, it is entirely possible to still trigger the issue on the rare case that a webtoon image has large horizontal dimension even after slicing them to display height as the PR proposes.

**Comment by CrepeTF** [comment]
> Would like to add that this issue also exists on **non-webtoon** images
> 
> **Steps to reproduce:**
> 
> 1. Install Mangadex extension.
> 2. Browse to [Isekai Monster Breeder](https://mangadex.org/title/0a89a48d-2e94-49a7-aa05-61b24e0f2f86/isekai-monster-breeder).
> 3. Open Chapter 45 scanlated by Emily Manga Translations.
> 4. Navigate to Page 2. Zoom in. This and every page after page 2 has very large dimensions and lags when panning. Even just bringing up the UI by tapping once anywhere in the middle is also laggy.
> 
> This is reproducible even on an LG G7. _Reading mode doesn't matter._
> 
> The solution in #7010 first proposed by #6973 doesn't fix this and the underlying problem. Furthermore, it is entirely possible to still trigger the issue on the rare case that a webtoon image has large horizontal dimension even after slicing them to display height as the PR proposes.

Can't reproduce

**Comment by X9VoiD** [comment]
My bad, will update the reproduction steps. "32-bit color" should be enabled to reproduce.

**Comment by CrepeTF** [comment]
@X9VoiD 

Works smoothly on my end, with 32-bit enabled
https://user-images.githubusercontent.com/70870719/165355086-678a6b3f-8923-42f3-b120-3f9b2edc3880.mp4

**Comment by X9VoiD** [comment]
Then I guess my phone is *slow*.

**Comment by ZaneHannanAU** [comment]
able to reproduce on ROG Phone 5, noting some stuttering even in high power mode. Longer comics (100k+ tall, 7-800px wide) do share this issue. Old Pixel shares it, but common slowdown is much quicker.

**Comment by samlux04** [comment]
Still no fix? Recently I read a lot of manhua and manhwa. I found a lot of lagginess in many chapters. I notice it's due to the very long images.
I can read it just fine and smooth with WebView/browser tho. Just not so pleasant due to many distraction around. And had to be careful when scrolling because it's not fixed width.

**Comment by Seishirou101** [comment]
there is currently a partial workaround in preview atm where it will split the images if they are downloaded. Though depending on your device it can also fail to split due to the huge image.

**Comment by Disposablegloves** [comment]
I can confirm that this occurs in non-webtoon stuff as well. Turning off 32-bit helped remove the frame rate drop for single pages, but wide two page spreads still suffer.

Here's a couple pages if you guys want to try. https://files.catbox.moe/fhvbyj.zip

If it matters, I was reading from a cbz file.

Phone is a Mi 10t Pro. 8 gb ram.

**Comment by kyoyacchi** [comment]
Same here. It's been 2 years....

**Comment by lemon07r** [comment]
> there is currently a partial workaround in preview atm where it will split the images if they are downloaded. Though depending on your device it can also fail to split due to the huge image.

How do we enable this?

**Comment by kyoyacchi** [comment]
> > there is currently a partial workaround in preview atm where it will split the images if they are downloaded. Though depending on your device it can also fail to split due to the huge image.
> 
> How do we enable this?

![Screenshot_20220911-104516](https://user-images.githubusercontent.com/63583961/189517390-1ced63ad-ee20-417b-a29a-950247cd9ebf.jpg)
Like this, i guess...

**Comment by lemon07r** [comment]
> > > there is currently a partial workaround in preview atm where it will split the images if they are downloaded. Though depending on your device it can also fail to split due to the huge image.
> > 
> > 
> > How do we enable this?
> 
> ![Screenshot_20220911-104516](https://user-images.githubusercontent.com/63583961/189517390-1ced63ad-ee20-417b-a29a-950247cd9ebf.jpg) Like this, i guess...

Found it! Thanks a ton, this was driving me mad. You're a lifesaver

**Comment by kayaoe** [comment]
Any fix for large Manga images causing lag for 120hz screens with 32-bit color enabled? I am using a Galaxy Tab S8 Plus.

Could the 32-bit color setting be toggled on for certain Manga, Manwha, webtoons, etc. and off for others with these settings being saved?

**Comment by Crayolable** [comment]
[I'm also experiencing this issue while scrolling through webtoons](https://youtu.be/W7jSz9fdL50) on my S23 Ultra. None of the following steps worked:
• Disabled "32-bit color"
• Enabled "Split tall images"
• Erased data and cache for Tachyomi 
• Uninstalled and then reinstalled Stable 0.14.5
• Changed the app's Battery setting to "Unrestricted"
